### PR TITLE
[lldb/formatter] Add Swift.UnsafeRawBufferPointer data formatter

### PR DESCRIPTION
**Pull request by medismailben**
This patch adds support for Swift's UnsafeRawBufferPointer and
UnsafeMutableRawBufferPointer data formatting.

This commit also introduces a new SwiftUnsafeType factory to adapt and
use both Unsafe{,Raw}BufferPointer with the pre-existing data formatter.

rdar://63946760

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

### CODE REVIEW

**Comment by medismailben** [comment]
@swift-ci test

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 1]
```diff
@@ -74,9 +77,20 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+  if (!start_addr || start_addr == LLDB_INVALID_ADDRESS)
+    return false;
+
+  m_start_addr = start_addr;
+
+  return true;
+}
+
+lldb::addr_t SwiftUnsafeBufferPointer::GetAddress(llvm::StringRef child_name) {
+  ConstString g_child_name(child_name);
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: g_something usually means the thing is a static, either local to the function or a file static.  This is neither, so g_ isn't an appropriate name.

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 2]
```diff
@@ -89,35 +103,104 @@ bool SwiftUnsafeBufferPointer::Update() {
   if (CompilerType type = some_value_sp->GetCompilerType())
     argument_type = SwiftASTContext::GetGenericArgumentType(type, 0);
 
-  if (!argument_type.IsValid())
-    return false;
-
-  m_elem_type = argument_type;
+  if (argument_type.IsValid())
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: Why is it okay to keep going here if you can't get a valid type for the contents of the UnsafeBufferPointer?  If there is a good reason, it would be good to explain it here.

**Comment by medismailben** [comment]
@swift-ci test Linux Platform

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 3]
```diff
@@ -149,7 +232,7 @@ class UnsafeBufferPointerSyntheticFrontEnd : public SyntheticChildrenFrontEnd {
   uint8_t m_ptr_size;
   lldb::ByteOrder m_order;
 
-  SwiftUnsafeBufferPointer m_unsafe_ptr;
+  SwiftUnsafeBufferPointer *m_unsafe_ptr;
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: You are leaking this SwiftUnsafeBufferPointer now.  Probably better to make this a unique_ptr<SwiftUnsafeBufferPointer>, then it will get cleaned up automatically.

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 4]
```diff
@@ -89,35 +103,104 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+  size_t count = 0;
+  addr_t addr = LLDB_INVALID_ADDRESS;
 
-  if (!swift_ubp.Update())
-    return false;
+  llvm::StringRef type_name(valobj.GetTypeName().GetCString());
 
-  size_t count = swift_ubp.GetCount();
-  addr_t addr = swift_ubp.GetStartAddress();
+  if (!type_name.contains("Raw")) {
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: Be nice to unify this.  If you had a factory that returned the right kind of object as a SwiftUnsafeBufferPointerUP then you wouldn't need the if/else

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 5]
```diff
@@ -170,13 +252,18 @@ lldb_private::formatters::swift::UnsafeBufferPointerSyntheticFrontEnd::
   m_ptr_size = process_sp->GetAddressByteSize();
   m_order = process_sp->GetByteOrder();
 
+  llvm::StringRef valobj_type_name(valobj_sp->GetTypeName().GetCString());
+  bool is_raw = valobj_type_name.contains("Raw");
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: It would be better to have a factory method that makes the right version as a UP.  Then you could set it into your ivar here and in the Summary you could also use that to avoid the ugly if (Raw) {make_raw; do stuff; } else {make_not_raw; do same stuff } construct.

File: lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift [hunk 6]
```diff
@@ -117,7 +117,43 @@ func main() {
...
+
+    mutrawbuf.swapAt(0, 1)
+    //% self.expect("frame variable -d run-target mutrawbuf",
+    //%            patterns=[
+    //%            '\(UnsafeMutableRawBufferPointer\) mutrawbuf = 2 values \(0[xX][0-9a-fA-F]+\) {',
+    //%            '\[0\] = 1',
+    //%            '\[1\] = 0',
+    //%            ])
+  }
+
```
Comment (jimingham) [inline lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:171]: Be good to test that "frame var mutrawbuf[0]" also works.

**Review by jimingham — COMMENTED** [review]
This looks pretty good.  The synthetic provider is leaking its SwiftUnsafeBufferPointer object, and I think you can isolate the choice to make the raw vrs. not raw difference.  Once you've made the right object only call base class methods so you should always handle the base class and let a factory make the right one for you.

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp (hunk 2 above)
Comment (medismailben) [reply lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: `UnsafeRawBufferPointer` doesn't have a generic argument since by design, it's a raw byte view of the pointed object. Since `GetAddress` is used for both type, when it's called from the `UnsafeRawBufferPointer` helper class, the generic argument type will be invalid, which leaves `m_elem_type` uninitialized (invalid). Later on, in `UnsafeRawBufferPointer::Update`, I check if `m_elem_type` is invalid, in which case, I reconstruct the `Swift.UInt8` compiler type.

If you look at `UnsafeRawBufferPointer`'s layout, we can't get the generic argument type with the top-level ValueObject ("_position"). To get it, lldb need to look one level below, at the "some" ValueObject node.
That's the reason why I do it in `GetAddress`, on the way of fetching the "_pointerValue" address.

I should probably add a comment to explain this.

File: lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift (hunk 6 above)
Comment (medismailben) [reply lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:171]: I'll add a check for that

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp (hunk 5 above)
Comment (medismailben) [reply lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: Will do!

**Comment by medismailben** [comment]
Thanks for the feedbacks @jimingham ! Indeed the factory approach would be a good way to manage both types. I'll wait to see if @dcci and @fredriss could think of other improvements I could make then update the PR.

**Comment by SNiTEBoBy** [comment]
Merge pull requested

**Comment by medismailben** [comment]
I updated the patch addressing @jimingham comments:

- Added a `SwiftUnsafeType` factory for UnsafeBufferPointer and UnsafeRawBufferPointer
- Used `std::unique_ptr` instead of a raw pointer.
- Added the test Jim requested.

**Comment by medismailben** [comment]
@swift-ci test

**Comment by medismailben** [comment]
@swift-ci test

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 7]
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+    return false;
+
+  m_count = (m_end_addr - m_start_addr) / *opt_type_size;
+
   return true;
 }
 
+std::unique_ptr<SwiftUnsafeType> SwiftUnsafeType::Create(ValueObject &valobj) {
+  llvm::StringRef valobj_type_name(valobj.GetTypeName().GetCString());
+  bool is_raw = valobj_type_name.contains("Raw");
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:305]: The formatter matching looks through typedefs.  So your valobj might have matched the formatter, but the top-level typename of this value object might be a typedef (and thus not contain "Raw").  The matching needs to also look through typedefs to the resolved type or you might get the wrong formatter.

**Review by jimingham — COMMENTED** [review]
That looks better to me.  

You have to handle the case where the valobj is a typedef to UnsafeBufferPointer or UnsafeRawBufferPointer.  Just strip off typedef's before looking for the "Raw" in the typename.

You might also want to think a bit about "if this were to fail in the field, where would I wish I had put in logging to help me figure out what went wrong" and if there are any obvious places add some "formatter" channel logging

**Review by SNiTEBoBy — APPROVED** [review]

File: lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift [hunk 8]
```diff
@@ -28,9 +28,9 @@ struct Plane: Flyable {
 }
 
 class Number<T:Numeric> {
-  
+
```
Comment (dcci) [inline lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift]: spurious.

File: lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift [hunk 9]
```diff
@@ -50,7 +50,7 @@ func main() {
     //%            '\[2\] = \(original = 3, opposite = -3\)',
     //%            ])
   }
-  
+
```
Comment (dcci) [inline lldb/test/API/functionalities/data-formatter/swift-unsafe/main.swift:52]: etc...

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 10]
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
...
 bool lldb_private::formatters::swift::UnsafeBufferPointerSummaryProvider(
     ValueObject &valobj, Stream &stream, const TypeSummaryOptions &options) {
 
-  SwiftUnsafeBufferPointer swift_ubp(valobj);
+  size_t count = 0;
+  addr_t addr = LLDB_INVALID_ADDRESS;
 
-  if (!swift_ubp.Update())
-    return false;
+  std::unique_ptr<SwiftUnsafeType> unsafe_type =
```
Comment (dcci) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: clang-format?

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 11]
```diff
@@ -74,50 +114,108 @@ bool SwiftUnsafeBufferPointer::Update() {
...
   m_start_addr = addr;
 
+  addr = GetAddress("_end");
+  if (!addr || addr == LLDB_INVALID_ADDRESS)
+    return false;
+  m_end_addr = addr;
+
+  if (!m_elem_type.IsValid()) {
+    auto ast_context = m_valobj.GetScratchSwiftASTContext();
+    if (!ast_context)
```
Comment (dcci) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: you need to grab the ASTContext lock & make sure the ASTContext has no errors.

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 12]
```diff
@@ -9,24 +9,64 @@
...
   CompilerType GetElementType() const { return m_elem_type; }
-  bool Update();
+  virtual bool Update() = 0;
 
-private:
+protected:
+  SwiftUnsafeType(ValueObject &valobj);
+  addr_t GetAddress(llvm::StringRef child_name);
+
+protected:
```
Comment (dcci) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: protected twice?

**Review by dcci — CHANGES_REQUESTED** [review]
A round of comments.

**Comment by medismailben** [comment]
@swift-ci test

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 13]
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+  //
+  // ▿ UnsafeRawBufferPointer
+  //   ▿ _position : Optional<UnsafeRawPointer<Int>>
+  //     ▿ some : UnsafeRawPointer<Int>
+  //       - pointerValue : Int
+  //   ▿ _end : Optional<UnsafeRawPointer<Int>>
+  //     ▿ some : UnsafeRawPointer<Int>
+  //       - pointerValue : Int
+  //
+  // The structure has 2 Optional UnsafePointer to the buffer's start address
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: UnsafePointer -> UnsafePointers

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 14]
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+  // and end address. To access the pointer address, lldb unfolds every value
+  // object child until reaching `pointerValue`.
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: value object -> ValueObject?

File: lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp [hunk 15]
```diff
@@ -74,50 +118,120 @@ bool SwiftUnsafeBufferPointer::Update() {
...
+  if (!type.IsValid())
+    return nullptr;
+
+  auto type_system = llvm::dyn_cast<TypeSystemSwift>(type.GetTypeSystem());
+  if (!type_system)
+    return nullptr;
+
+  opaque_compiler_type_t qual_type = type.GetOpaqueQualType();
+
+  if (type_system->IsTypedefType(qual_type))
```
Comment (jimingham) [inline lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp]: GetTypedefType only undoes one level of typedef'ing.  So the result might still be a typedef.  You probably need to loop over IsTypedefType till you get to the non-typedef'ed version.

File: lldb/source/Target/SwiftLanguageRuntimeNames.cpp [hunk 16]
```diff
@@ -393,13 +393,15 @@ SwiftLanguageRuntime::DemangleSymbolAsString(StringRef symbol, DemangleMode mode
     options.DisplayModuleNames = true;
     options.ShowPrivateDiscriminators = false;
     options.DisplayExtensionContexts = false;
+    options.DisplayLocalNameContexts = false;
```
Comment (jimingham) [inline lldb/source/Target/SwiftLanguageRuntimeNames.cpp:396]: Why was this necessary?

**Review by jimingham — COMMENTED** [review]
A few nits and one annoyance w.r.t. typedef's.

File: lldb/source/Target/SwiftLanguageRuntimeNames.cpp (hunk 16 above)
Comment (medismailben) [reply lldb/source/Target/SwiftLanguageRuntimeNames.cpp:396]: Previously, when a type alias was created inside a closure, the Swift demangler included the closure signature in the demangled type:

```
ByteBuffer #1 in closure #6 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> () in main.main() -> ()
```

In my case, this was causing an assertion when fetching the Typedef CompilerType, because both signature didn't match . This was fixed with #1331 and https://github.com/apple/swift/pull/32336.

Now, by setting `DemangleOptions::DisplayLocalNameContexts` to `false`, the demangler only returns the actual type (`ByteBuffer`), avoiding the assertion.

**Comment by medismailben** [comment]
Addressed final round of comments:
- Handled typedef chaining and added test for it
- Fixed typos.
- Added logging.

**Comment by medismailben** [comment]
@swift-ci test

**Comment by medismailben** [comment]
@swift-ci test
